metadata/src/main/java/org/apache/kafka/image/FeaturesImage.java

```diff
@@ -65,24 +64,43 @@ private Optional<Short> finalizedVersion(String feature) {
         return Optional.ofNullable(finalizedVersions.get(feature));
     }
 
-    public void write(Consumer<List<ApiMessageAndVersion>> out) {
-        List<ApiMessageAndVersion> batch = new ArrayList<>();
-        if (!metadataVersion.isLessThan(MetadataVersion.MINIMUM_BOOTSTRAP_VERSION)) {
-            // Write out the metadata.version record first, and then the rest of the finalized features
-            batch.add(new ApiMessageAndVersion(new FeatureLevelRecord().
-                    setName(MetadataVersion.FEATURE_NAME).
-                    setFeatureLevel(metadataVersion.featureLevel()), FEATURE_LEVEL_RECORD.lowestSupportedVersion()));
+    public void write(ImageWriter writer, ImageWriterOptions options) {
+        if (options.metadataVersion().isLessThan(MetadataVersion.MINIMUM_BOOTSTRAP_VERSION)) {
+            writePreProductionVersion(writer, options);
+        } else {
+            writeProductionVersion(writer, options);
         }
+    }
+
+    private void writePreProductionVersion(ImageWriter writer, ImageWriterOptions options) {
```

Comment:
Instead of `writePreProductionVersion` and `writeProductionVersion` can use a more focused name? How about `handleFeatureLevelNotSupported` and `writeFeatureLevels` respectively?